### PR TITLE
Fix base64 validation and add unittests

### PR DESCRIPTION
**Pull request by appgurueu**
Fixes base64 validation: Only up to 2 padding characters may appear at the end of base64.

### CODE REVIEW

File: src/util/base64.cpp [hunk 1]
```diff
@@ -33,16 +33,30 @@ static const std::string base64_chars =
 		"abcdefghijklmnopqrstuvwxyz"
 		"0123456789+/";
 
+static const std::string base64_chars_padding_1 = "AEIMQUYcgkosw048";
+static const std::string base64_chars_padding_2 = "AQgw";
 
 static inline bool is_base64(unsigned char c)
 {
-	return isalnum(c) || c == '+' || c == '/' || c == '=';
+	return isalnum(c) || c == '+' || c == '/';
```
Comment (numberZero) [inline src/util/base64.cpp:41]: ...yet `isalnum` is locale-dependent.

File: src/util/base64.cpp [hunk 2]
```diff
@@ -33,16 +33,30 @@ static const std::string base64_chars =
...
 {
-	for (char i : s)
-		if (!is_base64(i))
+	int i = 0;
+	for (; i < s.size(); ++i)
+		if (!is_base64(s[i]))
+			break;
+	// number of padding characters needs to match
+	unsigned char padding = 3 - ((s.size() + 3) % 4);
+	if (s.size() - i != padding
```
Comment (numberZero) [inline src/util/base64.cpp:52]: > Omission of padding characters is allowed

but where do you handle that?

Comment (appgurueu) [reply src/util/base64.cpp:52]: I forgot to, will need a == 0

File: src/util/base64.cpp (hunk 1 above)
Comment (appgurueu) [reply src/util/base64.cpp:41]: will change to explicit range checks

**Comment by SmallJoker** [comment]
> lhofhansl What's wrong with testing a valid string by checking for nil returned from decode?
> lhofhansl But no problem with it in concept.
> [...]
> Krock yes. that's my concern. IMO it does not need another function, if another one is already capable doing it for almost the same speed

https://irc.minetest.net/minetest-dev/2020-12-19#i_5766528

I like the corrected base64 tail detection, but would prefer to see that as a separate PR (bugfix) rather combined with a feature. Why exactly is this feature required? IIRC there was an open issue about this somewhere.

**Review by rubenwardy — CHANGES_REQUESTED** [review]
Please remove is_valid_base64 and only fix the validation. Users can check whether a string is valid base64 by checking for nil with decode_base64

**Comment by appgurueu** [comment]
> Please remove is_valid_base64 and only fix the validation. Users can check whether a string is valid base64 by checking for nil with decode_base64

Done.

File: doc/client_lua_api.txt [hunk 3]
```diff
@@ -907,7 +907,8 @@ Call these functions only at load time!
     * Example: `minetest.rgba(10, 20, 30, 40)`, returns `"#0A141E28"`
 * `minetest.encode_base64(string)`: returns string encoded in base64
     * Encodes a string in base64.
-* `minetest.decode_base64(string)`: returns string
+* `minetest.decode_base64(string)`: returns string or nil for invalid base64
+    * Padding needs to be omitted for 5.3.0 and lower not to return nil
```
Comment (sfan5) [inline doc/client_lua_api.txt:911]:
```suggestion
    * Before 5.5.0 this function did not accept padding and returned nil
```

File: doc/lua_api.txt [hunk 4]
```diff
@@ -5646,6 +5646,7 @@ Misc.
 * `minetest.encode_base64(string)`: returns string encoded in base64
     * Encodes a string in base64.
 * `minetest.decode_base64(string)`: returns string or nil for invalid base64
+    * Padding needs to be omitted for 5.3.0 and lower not to return nil
```
Comment (sfan5) [inline doc/lua_api.txt:5649]: (same)

File: src/util/base64.cpp [hunk 5]
```diff
@@ -33,16 +33,34 @@ static const std::string base64_chars =
...
-	for (char i : s)
-		if (!is_base64(i))
+	int i = 0;
+	for (; i < s.size(); ++i)
+		if (!is_base64(s[i]))
+			break;
+	// number of padding characters needs to match
+	unsigned char padding = 3 - ((s.size() + 3) % 4);
+	int actual_padding = s.size() - i;
+	if ((actual_padding != padding && actual_padding != 0)
```
Comment (sfan5) [inline src/util/base64.cpp:56]: shouldn't `actual_padding == 0` cause none of these checks to be ran?

Comment (appgurueu) [reply src/util/base64.cpp:56]: well, it doesn't matter. I could return early if that's what you want.

Comment (sfan5) [reply src/util/base64.cpp]: Maybe it already works correctly but I had trouble understanding the logic to verify its' correctness

**Comment by appgurueu** [comment]
The tests are not at all exhaustive, but provide a good starting point.

File: doc/lua_api.txt [hunk 6]
```diff
@@ -5646,6 +5646,7 @@ Misc.
 * `minetest.encode_base64(string)`: returns string encoded in base64
     * Encodes a string in base64.
 * `minetest.decode_base64(string)`: returns string or nil for invalid base64
+    * Before 5.5.0 this function did not accept padding and returned nil
```
Comment (SmallJoker) [inline doc/lua_api.txt:5649]: Really? Padding was necessary to get the skinsdb updater to work, hence commit f34abaedd2.

I believe this should state 5.4.0.

Comment (appgurueu) [reply doc/lua_api.txt:5649]: To be precise: Before 5.4 it did not accept padding and returned nil; 5.4 accepts invalid padding, which most likely results in undefined behavior.

Comment (SmallJoker) [reply doc/lua_api.txt:5649]: Right. That would fit better as a correect description.

**Review by SmallJoker — APPROVED** [review]

**Pull request merged**